# Importing Libraries

In [2]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# Importing Data

In [3]:
# Create Folder Path

path = r'/Users/queency_yustiawan/Study/CF/data analyst course/immersion/A4 exercise/Instacart Basket Analysis'

In [4]:
# Import products.csv data

df_prods = pd.read_csv (os.path.join (path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Import orders_wrangled.csv data

df_ords = pd.read_csv (os.path.join (path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data Consistency Check

In [30]:
df_ords.describe ()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  order_interval_days  
count        3.421083e+06       3.421083e+06         3.214874e+06  
mean         2.776219e+00       1.345202e+01         1.111484e+01  
std          2.046829e+00       4.226088e+00         9.206737e+00  
min          0.000000e+00       0.000000e+00         0.000000e+00  
25%          1.000000e+00       1.000000e+01         4.000000e+00  
50%          3.000000e+00       1.300000e+01         7.000000e+00  
75%          5.000000e+00       1.600000e+01         1.500000e+01  
max          6.000000e+00       2.300000e+01         3.000000e+01

## Mixed Type Data Exercise

In [35]:
# Create a Dataframe

df_test=pd.DataFrame()

# Create a Mixed Type Column

df_test['mix'] = ['a', 'b', 1, True]

df_test.head()

mix
0     a
1     b
2     1
3  True

In [37]:
# Check for mixed types

for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis =1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [38]:
# Converting the column's data to string

df_test['mix'] = df_test['mix'].astype('str')

## Missing Values Exercise

In [6]:
# Finding Missing Values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
# Create Subset for the Missing Value + checking what kind of value is missing

df_nan = df_prods [df_prods ['product_name'].isnull() == True]

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
# checking the total rows of original df_prods

df_prods.shape

(49693, 5)

In [10]:
# creating a new subset contain only non-missing value (due to missing values are strings data, can't be imputed)

df_prods_clean = df_prods [df_prods['product_name'].isnull() == False]

In [11]:
# checking the total rows of new subset 

df_prods_clean.shape

(49677, 5)

# Duplicate Data Exercise

In [12]:
# Finding the duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [13]:
# Creating new dataframe without the duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [15]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting Cleaned Products dataframe

In [16]:
# Exporting df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv (os.path.join (path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Task 4.5

## 2. Run df.describe () on df_ords and share the inconsistencies

In [17]:
# Checking df_ords inconsistencies


df_ords.describe ()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  order_interval_days  
count        3.421083e+06       3.421083e+06         3.214874e+06  
mean         2.776219e+00       1.345202e+01         1.111484e+01  
std          2.046829e+00       4.226088e+00         9.206737e+00  
min          0.000000e+00       0.000000e+00         0.000000e+00  
25%          1.000000e+00       1.000000e+01         4.000000e+00  
50%          3.000000e+00       1.300000e+01         7.000000e+00  
75%          5.000000e+00       1.600000e+01         1.500000e+01  
max          6.000000e+00       2.300000e+01         3.000000e+01

####  "order_number" min = 1, max = 100. mean = 17, median = 11
####  "orders_of_day_of week" min = 0, max = 6. mean = 2.78, median = 3 
####  "order_hour_of_day" min = 0, max = 23, mean = 13.5, median = 13
####  above columns are all good

#### The count in "order_interval_days" is lower compared to others, showing a potential of missing values. Need to investigate further
#### min = 0, max = 30, mean = 11, median = 7


# 3. Check for mixed type data in df_ords

In [24]:
# Checking for mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis =1)
    if len (df_ords[weird]) > 0:
        print(col)
    else:
        print ("None of the columns have mix data")

None of the columns have mix data
None of the columns have mix data
None of the columns have mix data
None of the columns have mix data
None of the columns have mix data
None of the columns have mix data
None of the columns have mix data


# 5. Check for missing values in df_ords

In [25]:
# Checking for missing values

df_ords.isnull().sum()

Unnamed: 0                  0
order_id                    0
user_id                     0
order_number                0
orders_day_of_week          0
order_hour_of_day           0
order_interval_days    206209
dtype: int64

#### There are 206,209 values missing in the "order_interval_days" column. It is quite a significant volume of data so it's not safe to be removed. We should first investigate further if there is a possibility to impute the missing value, or, otherwise we should create a new subset dataframe without the missing value.

# 6. Address the missing value

In [28]:
# Creating a subset containing only the missing value

df_ords_nan = df_ords [df_ords['order_interval_days'].isnull() == True]

df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  order_interval_days  
0                        8                  NaN  
11                      11                  NaN  
26                      14                  NaN  
39                      11                  NaN  
45                      12                  NaN  
...                    ...                  ...  
3420930                 12                  NaN  
3420934                 18                  NaN  
3421002                 19                  NaN  
3421019                 15                  NaN  
3421069                 11                  NaN  

[206209 rows x 7 columns]

#### All the order number in the missing value of "order_interval_days" seems to be 1, so it is possible that "NaN" is created for every first time order of the customer.

In [30]:
# Checking the frequency of order_number 1

df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

#### the total count of order_number 1 is 206,209 which is the same as the missing values count. As it is impossible to impute this data, we should leave it untouched. It is also not ideal to create a subset dataframe without the missing value, as it will be removing all of the customers' first order from the dataframe. 

# 7. Check duplicate values for df_ords

In [32]:
# Checking duplicates in df_ords

df_ords_dups = df_ords [df_ords.duplicated()]

df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, order_interval_days]
Index: []

# 8. Address the duplicate data

#### There are no duplicate data in df_ords as shown above in df_ords_dups. It has no value for all columns.

# 9. Export the data

In [33]:
# Exporting df_ords

df_ords.to_csv (os.path.join (path, '02 Data', 'Prepared Data', 'orders_checked.csv'))